src/NuGet.Core/NuGet.Packaging/Signing/Archive/ReadOnlyBufferedStream.cs

```diff
@@ -0,0 +1,365 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.Globalization;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace NuGet.Packaging.Signing
+{
+    public sealed class ReadOnlyBufferedStream : Stream
+    {
+        private const int _defaultBufferSize = 4096;
+
+        private readonly byte[] _buffer;
+        private readonly bool _leaveOpen;
+        private readonly Lazy<long> _length;
+        private readonly Stream _stream;
+
+        private long _bufferStartPosition;
+        private int _bufferFillLength;
+        private bool _isDisposed;
+        private long _position;
+
+        public override bool CanRead => _stream.CanRead;
+        public override bool CanSeek => _stream.CanSeek;
+        public override bool CanTimeout => _stream.CanTimeout;
+        public override bool CanWrite => false;
+        public override long Length => _length.Value;
+
+        public override long Position
+        {
+            get
+            {
+                ThrowIfDisposed();
+
+                return _position;
+            }
+            set
+            {
+                ThrowIfDisposed();
+
+                if (value < 0)
```

Comment:
Should we also check that value is not bigger than length?

Reply:
No, that's not the contract of `Stream`.  There are streams that support seeking and streams that do not.  Of the streams that do, ["[s]eeking to any location beyond the length of the stream is supported."](https://msdn.microsoft.com/en-us/library/system.io.stream.position(v=vs.110).aspx#Anchor_2)

For example, `FileStream`, when opened in read/write mode, allows you to [grow a file by seeking beyond the end of the stream and then writing](https://msdn.microsoft.com/en-us/library/system.io.filestream.position(v=vs.110).aspx#Anchor_2).

The behavior of this stream is consistent with `MemoryStream` in read-only mode.

                using (var stream = new MemoryStream(new byte[1], writable: false))
                {
                    // Seek beyond the end of the stream.
                    stream.Position = 5;

                    // ReadByte() returns -1.
                    var value = stream.ReadByte();

                    // Read(...) returns 0.
                    var bytesRead = stream.Read(new byte[1], offset: 0, count: 1);
                }